ui: add top-level LoginContainer: require login before rendering anything

vilterp opened the pull request:
Depends on #25057 
Touches #24939

![](https://user-images.githubusercontent.com/7341/39150096-38b4cd28-470f-11e8-9d67-e1832d35a211.gif)

Shown above:
1. go to admin UI; see login screen
2. error message when you type in the wrong password
3. you can't hit an authenticated endpoint because you don't have a valid session (this checking is turned on by #24944, only in secure mode)
4. once you login with the right password, you can see the UI (the temporary "connection lost" banner shouldn't be there)
5. now you (and the UI itself) can hit endpoints, because you have a valid session

Todos:
- [ ] redirect to login page instead of current wrapper component switching thing
- [ ] logout
    - [ ] logout button (make API call; redirect to login & blow away redux store)
    - [ ] log out other tabs (if they get a 403, redirect to login & blow away redux store)
- [ ] styling

Release note: None

Commit:
ui: add top-level LoginContainer: require login before rendering anything

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/25005)
<!-- Reviewable:end -->

vilterp commented on pkg/ui/src/views/login/loginContainer.tsx:
```diff
@@ -0,0 +1,40 @@
+import Layout from "src/views/app/containers/layout";
+
+interface LoginContainerState {
+  loggedIn: boolean;
```
note: move this into redux store

mrtracy commented on pkg/ui/src/views/login/loginContainer.tsx:
```diff
@@ -0,0 +1,40 @@
+  constructor(props: RouterState) {
+    super(props);
+    // TODO(vilterp): the user might already have a valid session. how do we check that?
+    // maybe do some templating on the backend and add it in.
```
Local storage might be an answer, with a fallback on authorization failure.  A quick outline:

1. This needs to be in the redux state anyway - move this there.
2. After creating the store for the actual page, initialize it with one-time code that checks local storaage and, if it finds the logged-in flat, updates the redux state before the initial rendering.
3. When the login method actually succeeds, write to this local state after updating the redux state.
4. If a user method gets an authorization error, we need to clear this logged-in state if it exists.

You might see if there is some blessed method of synchronizing pieces of the redux state with local storage, but I doubt it's especially tricky.  #4 seems like the hard part.

vilterp replied:
Didn't think of local storage, but it makes sense. My plan had been to make `index.html` a Go template into which I would dump `window.loggedInUser = {{ loggedInUser }}` or similar. However, `index.html` is not a nice little static file I can make into a template; it's generated through webpack (ultimately from this: https://github.com/jaketrent/html-webpack-template/blob/master/index.ejs)

So, local storage could work nicely. Will probably move forward with that unless @couchand has objections.

mrtracy commented on pkg/ui/src/views/login/loginContainer.tsx:
```diff
@@ -0,0 +1,40 @@
+  render() {
+    if (this.state.loggedIn) {
+      return (
+        <Layout {...this.props}>
```
The general idea here is sound to me; with this method, we also don't need to remember any query params or URL paths, since this just renders in place regardless of the path.  I think it's a good place to start.

vilterp replied:
@couchand said IRL he thinks this should redirect to `/login`, not just display the login page on whatever URL you're trying to go to. I think the argument is that with the redirect, we see `/login` in metrics.

Then we _do_ have to remember where you're trying to go somehow, with a GET parameter or something. (maybe it can be stuffed in the state of `LoginContainer` somehow, if that's still there).

vilterp commented:
K, rewrote w/ Redux. This also now depends on #25171, which involves setting `window.loggedInUser` in a script tag in index.html, generated client-side in a Go template. (PR coming soon)

couchand commented:
Reviewed 7 of 7 files at r6.
Review status: 7 of 11 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/ui/src/redux/login.ts, line 14 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/25005#-LBRdYdd4Cl60exgLjSY:-LBRdYdd4Cl60exgLjSZ:bqu9rcr) ([raw file](https://github.com/cockroachdb/cockroach/blob/85000c2bb97fcc55ee97da5adc9c386230f8260e/pkg/ui/src/redux/login.ts#L14)):*
> ```XML
> declare global {
>   interface Window {
>     loggedInUser: string;
> ```

probably `loggedInUser?`

---

*[pkg/ui/src/redux/login.ts, line 19 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/25005#-LBRdcbGBuDSo68oY0MP:-LBRdcbGBuDSo68oY0MQ:b-gkzdj4) ([raw file](https://github.com/cockroachdb/cockroach/blob/85000c2bb97fcc55ee97da5adc9c386230f8260e/pkg/ui/src/redux/login.ts#L19)):*
> ```XML
> 
> export interface LoginState {
>   loggedInUser: string;
> ```

likewise

---

*[pkg/ui/src/redux/login.ts, line 21 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/25005#-LBRdrRzAxnd-reV6ifg:-LBRdrRzAxnd-reV6ifh:bxwn5zd) ([raw file](https://github.com/cockroachdb/cockroach/blob/85000c2bb97fcc55ee97da5adc9c386230f8260e/pkg/ui/src/redux/login.ts#L21)):*
> ```XML
>   loggedInUser: string;
>   error: string;
>   inProgress: boolean;
> ```

suggest keeping the names of these props consistent with `CachedDataReducer`

---

*[pkg/ui/src/redux/login.ts, line 99 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/25005#-LBRe9DhD6TBxo45W2U_:-LBRe9DhD6TBxo45W2Ua:b59puxs) ([raw file](https://github.com/cockroachdb/cockroach/blob/85000c2bb97fcc55ee97da5adc9c386230f8260e/pkg/ui/src/redux/login.ts#L99)):*
> ```XML
>         dispatch(loginSuccess(username));
>       })
>       .catch((err) => {
> ```

suggest passing two functions to `then` to avoid catching anything thrown in the `dispatch(loginSuccess...`

---

*[pkg/ui/src/views/login/loginContainer.tsx, line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25005#-LAo3rnV2CT8JK6BpghL-r1-30:-LBRczsQ8-lX0NZhRjle:b3s0tio) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c6dd3e52b06b74cf9f1477a1af28850475f7b64/pkg/ui/src/views/login/loginContainer.tsx#L30)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

@couchand said IRL he thinks this should redirect to `/login`, not just display the login page on whatever URL you're trying to go to. I think the argument is that with the redirect, we see `/login` in metrics.

Then we _do_ have to remember where you're trying to go somehow, with a GET parameter or something. (maybe it can be stuffed in the state of `LoginContainer` somehow, if that's still there).
</blockquote></details>

The usual method is a query parameter with the URL to return to.  Need to consider the impact of URL redactions as well.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25005)*
<!-- Sent from Reviewable.io -->

vilterp commented:
Added a login indicator on the side. Says "insecure mode" if login is not enabled. (currently, enabled is `secure && COCKROACH_EXPERIMENTAL_REQUIRE_WEB_LOGIN env var`; see backend PR)

![image](https://user-images.githubusercontent.com/7341/39550635-fb8b42f6-4e2f-11e8-9ea0-c5bc26c8f585.png)

couchand committed:
ui: refactor LoginState to manage enabled/disabled

This allows us to have one object representing both "is login enabled?"
and "does this user have access?" state questions.

Release note: None

couchand committed:
ui: redirect to login page when not authned

Previously, the login page was swapped in for the content of a page if
a user was not logged in.  This changes the behavior to instead redirect
to a free standing login page.

The original URL accessed is stashed in the query parameter redirectTo,
and after authentication the user is returned to that page.

Release note: None.

couchand commented:
Updated to redirect.  Things remaining here:

- redacting the redirect url parameter,
- addressing inline feedback above
- logout & styling

vilterp commented:
Tested redirect; works nicely! Working on logout.

couchand committed:
ui: redact URLs encoded in query parameters

Prevent URLs from slipping though the analytics redactions if the are
encoded in a query parameter, as with the login redirect param.

Release note: None

vilterp commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Logout coming up next.

---

Reviewed 1 of 7 files at r6, 4 of 8 files at r7, 3 of 5 files at r8, 6 of 6 files at r9, 4 of 4 files at r10.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/ui/src/redux/login.ts, line 14 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/25005#-LBRdYdd4Cl60exgLjSY:-LD7v21bEv4-Ahh4VPl7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/85000c2bb97fcc55ee97da5adc9c386230f8260e/pkg/ui/src/redux/login.ts#L14)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

probably `loggedInUser?`
</blockquote></details>

Done.

---

*[pkg/ui/src/redux/login.ts, line 19 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/25005#-LBRdcbGBuDSo68oY0MP:-LD7v2cvC4e4dDhUQZ2g:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/85000c2bb97fcc55ee97da5adc9c386230f8260e/pkg/ui/src/redux/login.ts#L19)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

likewise
</blockquote></details>

Done.

---

*[pkg/ui/src/redux/login.ts, line 21 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/25005#-LBRdrRzAxnd-reV6ifg:-LD7vSHnA2XTJ2XW9sPH:b-ckdsvm) ([raw file](https://github.com/cockroachdb/cockroach/blob/85000c2bb97fcc55ee97da5adc9c386230f8260e/pkg/ui/src/redux/login.ts#L21)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

suggest keeping the names of these props consistent with `CachedDataReducer`
</blockquote></details>

I guess this would change to `inFlight`? (I don't really care tho)

---

*[pkg/ui/src/redux/login.ts, line 99 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/25005#-LBRe9DhD6TBxo45W2U_:-LD7vLJJ2QRzvlWT_R60:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/85000c2bb97fcc55ee97da5adc9c386230f8260e/pkg/ui/src/redux/login.ts#L99)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

Oh good idea, I didn't know you could do that. I think we do this wrong other places (e.g. CachedDataReducer), resulting in swallowing JS errors as if they're API call errors
</blockquote></details>

Done.

---

*[pkg/ui/src/views/login/loginContainer.tsx, line 8 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25005#-LAo3rnV2CT8JK6BpghL-r1-8:-LD7v6ZT3yx3TZasXgna:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c6dd3e52b06b74cf9f1477a1af28850475f7b64/pkg/ui/src/views/login/loginContainer.tsx#L8)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

TODO: move this into redux store
</blockquote></details>

Done.

---

*[pkg/ui/src/views/login/loginContainer.tsx, line 15 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25005#-LAo3rnV2CT8JK6BpghL-r1-15:-LD7vAen3-ykLDv1Wk9_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c6dd3e52b06b74cf9f1477a1af28850475f7b64/pkg/ui/src/views/login/loginContainer.tsx#L15)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

Didn't think of local storage, but it makes sense. My plan had been to make `index.html` a Go template into which I would dump `window.loggedInUser = {{ loggedInUser }}` or similar. However, `index.html` is not a nice little static file I can make into a template; it's generated through webpack (ultimately from this: https://github.com/jaketrent/html-webpack-template/blob/master/index.ejs)

So, local storage could work nicely. Will probably move forward with that unless @couchand has objections.
</blockquote></details>

Done.

---

*[pkg/ui/src/views/login/loginContainer.tsx, line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25005#-LAo3rnV2CT8JK6BpghL-r1-30:-LD7vAI23gUnnz0gKY-V:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c6dd3e52b06b74cf9f1477a1af28850475f7b64/pkg/ui/src/views/login/loginContainer.tsx#L30)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

Hm yes, I guess we'd need to make sure to run the value of the query param through redaction before sending to segment.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25005#-:-LD7vUvkApg4_FlRI5yH:b67dez2)*
<!-- Sent from Reviewable.io -->

couchand commented:
bors r+
